modules/scalacheck/shared/src/main/scala/weaver/scalacheck/CheckConfig.scala

```diff
@@ -1,18 +1,106 @@
+    maximumGeneratorSize = this.maximumGeneratorSize,
+    perPropertyParallelism = this.perPropertyParallelism,
+    initialSeed = this.initialSeed
+  )
+
+  def withMaximumDiscardRatio(maximumDiscardRatio: Int) = new CheckConfig(
+    minimumSuccessful = this.minimumSuccessful,
+    maximumDiscardRatio = maximumDiscardRatio,
+    maximumGeneratorSize = this.maximumGeneratorSize,
+    perPropertyParallelism = this.perPropertyParallelism,
+    initialSeed = this.initialSeed
+  )
+
+  def withMaximumGeneratorSize(maximumGeneratorSize: Int) = new CheckConfig(
+    minimumSuccessful = this.minimumSuccessful,
+    maximumDiscardRatio = this.maximumDiscardRatio,
+    maximumGeneratorSize = maximumGeneratorSize,
+    perPropertyParallelism = this.perPropertyParallelism,
+    initialSeed = this.initialSeed
+  )
+
+  def withPerPropertyParallelism(perPropertyParallelism: Int) = new CheckConfig(
+    minimumSuccessful = this.minimumSuccessful,
+    maximumDiscardRatio = this.maximumDiscardRatio,
+    maximumGeneratorSize = this.maximumGeneratorSize,
+    perPropertyParallelism = perPropertyParallelism,
+    initialSeed = this.initialSeed
+  )
+
+  def withInitialSeed(initialSeed: Option[Seed]) = new CheckConfig(
+    minimumSuccessful = this.minimumSuccessful,
+    maximumDiscardRatio = this.maximumDiscardRatio,
+    maximumGeneratorSize = this.maximumGeneratorSize,
+    perPropertyParallelism = this.perPropertyParallelism,
+    initialSeed = initialSeed
+  )
+
+  override def toString: String = {
+    val b = new StringBuilder("CheckConfig(")
+    b.append(String.valueOf(this.minimumSuccessful))
+    b.append(", ")
+    b.append(String.valueOf(this.maximumDiscardRatio))
+    b.append(", ")
+    b.append(String.valueOf(this.maximumGeneratorSize))
+    b.append(", ")
+    b.append(String.valueOf(this.perPropertyParallelism))
+    b.append(", ")
+    b.append(String.valueOf(this.initialSeed))
+    b.append(")")
+    b.toString()
+  }
+
+  override def canEqual(obj: Any): Boolean =
+    obj != null && obj.isInstanceOf[CheckConfig]
+
+  override def equals(obj: Any): Boolean = canEqual(obj) && {
+    val other = obj.asInstanceOf[CheckConfig]
+    this.minimumSuccessful == other.minimumSuccessful && this.maximumDiscardRatio == other.maximumDiscardRatio && this.maximumGeneratorSize == other.maximumGeneratorSize &&
+    this.perPropertyParallelism == other.perPropertyParallelism && this.initialSeed == other.initialSeed
+  }
+
+  override def hashCode: Int = {
+    var code = 17 + "CheckConfig".##
+    code = 37 * code + this.minimumSuccessful.##
+    code = 37 * code + this.maximumDiscardRatio.##
+    code = 37 * code + this.maximumGeneratorSize.##
+    code = 37 * code + this.perPropertyParallelism.##
+    code = 37 * code + this.initialSeed.##
+    37 * code
+  }
+
+  override def productPrefix: String = "CheckConfig"
+
+  override def productArity: Int = 5
+
+  override def productElement(n: Int): Any = n match {
```

Comment:
mmm, I had not remembered the implementation on the PR you merged in, not a fan of how it was done here. 

It's fine to use a case-class implementation (to get toString/equals/hashCode), as long as the constructor is made private (to hide copy methods), the default extractor is overriden as described [here](https://docs.scala-lang.org/overviews/core/binary-compatibility-for-library-authors.html#changing-a-case-class-definition-in-a-backwards-compatible-manner), and an explicit `apply` method is provided in the companion. It's less error-prone.

Reply:
Thanks! That article is incredibly valuable.